microsoft/testsuites/cvm/cvm_boot.py

```diff
@@ -0,0 +1,176 @@
+
+        Steps:
+        1. On first boot, check current PCR values for PCR4 and PCR7
+        2. Get current boot components versions (e.g. shim, grub, systemd-boot, uki)
+        3. Run a package upgrade to update boot components
+        4. Get new boot components versions to see if anything has changed
+        5. Reboot the CVM, make sure the CVM can boot up again
+        6. PCR4 should change if any of the boot components is upgraded
+        7. PCR7 may change (for example, if a signing certificate is changed)
+        """,
+        priority=1,
+        requirement=simple_requirement(
+            supported_features=[CvmEnabled()],
+            supported_platform_type=[AZURE],
+        ),
+    )
+    def verify_boot_success_after_component_upgrade(
+        self,
+        log: Logger,
+        node: RemoteNode,
+        environment: Environment,
+        log_path: Path,
+        result: TestResult,
+        **kwargs: Any,
+    ) -> None:
+        posix_os: Posix = cast(Posix, node.os)
+        # First boot
+        # - Check PCR values (PCR4, PCR7)
+        pcrs_before_reboot = node.tools[Tpm2].pcrread(pcrs=[4, 7])
+
+        # - Get current boot components versions (shim, systemd-boot, kernel-uki)
+        boot_components = ["shim", "systemd-boot", "kernel-uki"]
+        boot_components_versions = dict()
+        for pkg in boot_components:
+            pkg_info = posix_os.query_package(pkg)
+            boot_components_versions[pkg] = pkg_info.version_str
+
+        # - Upgrade boot components
+        posix_os.update_packages(boot_components)
+
+        # - Get new boot components versions
+        boot_components_new_versions = dict()
+        for pkg in boot_components:
+            pkg_info = posix_os.query_package(pkg)
+            boot_components_new_versions[pkg] = pkg_info.version_str
+
+        # Reboot
+        # - Make sure VM boots up again
+        reboot_tool = node.tools[Reboot]
+        reboot_tool.reboot_and_check_panic(log_path)
+        is_ready, tcp_error_code = wait_tcp_port_ready(
+            node.connection_info[constants.ENVIRONMENTS_NODES_REMOTE_ADDRESS],
+            node.connection_info[constants.ENVIRONMENTS_NODES_REMOTE_PORT],
+            log=log,
+        )
+        if not is_ready:
+            raise TcpConnectionException(
+                node.connection_info[constants.ENVIRONMENTS_NODES_REMOTE_ADDRESS],
+                node.connection_info[constants.ENVIRONMENTS_NODES_REMOTE_PORT],
+                tcp_error_code,
+                "no panic found in serial log",
+            )
+
+        pcrs_after_reboot = node.tools[Tpm2].pcrread(pcrs=[4, 7])
+        boot_component_changed = any(
+            boot_components_versions[pkg] != boot_components_new_versions[pkg]
+            for pkg in boot_components
+        )
+
+        # - PCR4 should change if any of the boot components is upgraded
+        # - PCR7 may change if a signing cert is changed
+        if boot_component_changed:
```

Comment:
Is it possible to make sure all paths tested in the same test case?

Reply:
I think this logic already covers what we want to test for CVM booting, at least for Azure Linux. If you're referring to merging the 2 test cases defined in this file, I think it's possible. If that's preferred then I'll update my code

Reply:
For the `if` logic, only one path (`if` or `else`) can be covered each time. From your logic, both `if/else` has assertion to validate something. Only one of them can be covered in a test pass, so another path is failed to cover.

Reply:
I see. We plan to run this test case against an Azure Linux CVM image on a per-release basis, so it's expected that we only hit one branch for each test run. From our side that is acceptable.

Reply:
I mean it will vary for each release. If we bump versions of any boot component package for a new release, the `boot_component_changed = True` case will run, and I expect this case to be run more frequently. The assertion in the other case is more like a sanity check and is _**very**_ unlikely to fail. I can remove that but I'd prefer to leave it there for completeness